compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -304,7 +307,8 @@ object Implicits:
      */
     override val level: Int =
       if outerImplicits == null then 1
-      else if migrateTo3(using irefCtx)
+      else if isImport && (irefCtx.owner eq outerImplicits.irefCtx.owner)
```

Comment:
Bikeshed: If desired, readability could be slightly improved by refactoring this boolean expression to:
```
      inline def isSameOwner = irefCtx.owner eq outerImplicits.irefCtx.owner
      inline def isSameScope = irefCtx.scope eq outerImplicits.irefCtx.scope
      inline def isLazyImplicit = refs.head.implicitName.is(LazyImplicitName)

      if outerImplicits == null then 1
      else if migrateTo3(using irefCtx)
              || isSameOwner && (isImport || isSameScope && !isLazyImplicit)
      then outerImplicits.level
      else outerImplicits.level + 1
```
 This way imports also might look a little bit less special than now.

Reply:
true. And I don't think we need the `inline`; the JVM is good at inlining these anyway.